aten/src/ATen/native/mkldnn/Conv.cpp

```diff
@@ -66,377 +98,362 @@ static std::vector<int64_t> conv_output_size(
-  auto input_memory = input_usr_memory;
-  if (input_usr_memory.get_primitive_desc() != memory::primitive_desc(input_pd)) {
-    input_memory = memory(input_pd);
-    net.push_back(reorder(input_usr_memory, input_memory));
-  }
+  auto grad_input = at::empty(input_size, grad_output.options());
+  ConvolutionArgs args(grad_input, weight, grad_output, padding, output_padding,
+      stride, dilation, groups, bias_defined, true);
 
-  auto grad_output_pd = conv_backward_weight_pd->diff_dst_primitive_desc();
-  auto grad_output_memory = grad_output_usr_memory;
-  if (grad_output_usr_memory.get_primitive_desc() != memory::primitive_desc(grad_output_pd)) {
-    grad_output_memory = memory(grad_output_pd);
-    net.push_back(reorder(grad_output_usr_memory, grad_output_memory));
-  }
+  auto dtype = get_mkldnn_dtype(grad_output);
+  desc grady_desc(args.output_tz, dtype);
+  desc weight_desc(args.weight_tz, dtype);
 
-  auto grad_weight_pd = conv_backward_weight_pd->diff_weights_primitive_desc();
-  auto grad_weight_memory = grad_weight_usr_memory;
-  if (grad_weight_usr_memory.get_primitive_desc() != memory::primitive_desc(grad_weight_pd)) {
-    grad_weight_memory = memory(grad_weight_pd);
-  }
+  itensor grady_, weight_, gradx_, gradx/*user layout*/;
+
+  grady_.init(grady_desc, grad_output.data_ptr());
+  weight_.init(weight_desc, weight.data_ptr());
+
+  desc gradx_desc(args.input_tz, grady_.get_data_type());
+  gradx.init(gradx_desc, grad_input.data_ptr());
+
+  convolution_transpose_backward_data::compute(grady_, weight_, args.input_tz,
+      gradx_, args._stride, args._padding, args._padding_r);
 
-  std::shared_ptr<convolution_backward_weights> conv_backward_weight;
+  reorder::compute(gradx_, gradx);
+
+  return grad_input;
+}
+
+std::tuple<Tensor, Tensor> mkldnn_convolution_transpose_backward_weights(
+    IntArrayRef weight_size, const Tensor& grad_output, const Tensor& input,
+    IntArrayRef padding, IntArrayRef output_padding, IntArrayRef stride,
+    IntArrayRef dilation, int64_t groups, bool bias_defined) {
+
+  auto grad_weight = at::empty(weight_size, grad_output.options());
+  Tensor grad_bias;
   if (bias_defined) {
-    grad_bias_memory.reset(new memory({{{bias_tz}, data_t, format_x}, cpu_engine},
-      grad_bias.data_ptr()));
-    conv_backward_weight.reset(new convolution_backward_weights(*conv_backward_weight_pd,
-      input_memory, grad_output_memory, grad_weight_memory, *grad_bias_memory));
-  } else {
-    conv_backward_weight.reset(new convolution_backward_weights(*conv_backward_weight_pd,
-      input_memory, grad_output_memory, grad_weight_memory));
+    grad_bias = at::empty({grad_output.size(1)}, grad_output.options());
   }
 
-  net.push_back(*conv_backward_weight);
+  ConvolutionArgs args(input, grad_weight, grad_output, padding,
+      output_padding, stride, dilation, groups, bias_defined, true);
+
+  auto dtype = get_mkldnn_dtype(input);
+  desc x_desc(args.input_tz, dtype);
```

Comment:
Could we name the descriptors consistently with the input names? E.g., so `input_desc` rather than `x_desc`

Reply:
(Unless, perhaps, these are the mkldnn names? In that case, I guess it's OK.)